utils/python/CIME/nmlgen.py

```diff
@@ -0,0 +1,610 @@
+"""Class for generating component namelists."""
+
+# Typically ignore this.
+# pylint: disable=invalid-name
+
+# Disable these because this is our standard setup
+# pylint: disable=wildcard-import,unused-wildcard-import
+
+import datetime
+import re
+
+from CIME.XML.standard_module_setup import *
+from CIME.namelist import Namelist, parse, \
+    character_literal_to_string, string_to_character_literal, \
+    expand_literal_list, compress_literal_list, merge_literal_lists
+from CIME.XML.namelist_definition import NamelistDefinition
+from CIME.utils import expect
+
+logger = logging.getLogger(__name__)
+
+_var_ref_re = re.compile(r"\$(\{)?(?P<name>\w+)(?(1)\})")
+
+_ymd_re = re.compile(r"%(?P<digits>[1-9][0-9]*)?y(?P<month>m(?P<day>d)?)?")
+
+_stream_file_template = """
+<dataSource>
```

Comment:
This is some psuedo xml code for generating stream files, the plan is to eventually get rid of this style of input and put streams into namelist files.